src/heaps/minmax_heap.jl

```diff
@@ -0,0 +1,261 @@
+#
+# utilities
+#
+################################################
+
+@inline level(i) = round(Int, floor(log2(i)))
+@inline lchild(i) = 2*i
+@inline rchild(i) = 2*i+1
+@inline children(i) = [lchild(i), rchild(i)]
+@inline hparent(i) = round(Int, floor(i/2))
+
+"""
+Return the indices of all children and grandchildren of
+position `i`.
+"""
+function descendants(N::T, i::T) where {T <: Integer}
+    _descendants = T[]
+    for child in children(i)
+        append!(_descendants, [child, lchild(child), rchild(child)])
+    end
+    return [d for d in _descendants if d ≤ N]
+end
+
+"""
+    is_minmax_heap(h::AbstractVector) -> Bool
+
+Return `true` if `A` is a min-max heap. A min-max heap is a
+heap where the minimum element is the root and the maximum
+element is a child of the root.
+"""
+function is_minmax_heap(A::AbstractVector)
+
+    isheap = true
+    N = length(A)
+
+    for i = 1:N
+        if level(i)%2 == 0
+            # min layer
+            # check that A[i] < children A[i]
+            #    and grandchildren A[i]
+            for j in descendants(N, i)
+                isheap &= A[i] ≤ A[j]
+            end
+        else
+            # max layer
+            for j in descendants(N, i)
+                isheap &= A[i] ≥ A[j]
+            end
+        end
+    end
+    isheap
+end
+
+################################################
+#
+# interfaces
+#
+################################################
+
+length(h::BinaryMinMaxHeap) = length(h.valtree)
+
+isempty(h::BinaryMinMaxHeap) = isempty(h.valtree)
+
+"""
+    pop!(h::BinaryMinMaxHeap) = popmin!(h)
+"""
+@inline pop!(h::BinaryMinMaxHeap) = popmin!(h)
+
+"""
+    popmin!(h::BinaryMinMaxHeap) -> min
+                        
+Remove the minimum value from the heap.
+"""
+function popmin!(h::BinaryMinMaxHeap)
+    valtree = h.valtree
+    x = valtree[1]
+    y = pop!(valtree)
+    if !isempty(valtree)
+        valtree[1] = y
+        _minmax_heap_trickle_down!(valtree, 1)
+    end
+    return x
+end
+
+"""
+    popmax!(h::BinaryMinMaxHeap) -> max
+                        
+Remove the maximum value from the heap.
+"""
+function popmax!(h::BinaryMinMaxHeap)
+    valtree = h.valtree
+    @inbounds x, i = maximum([(valtree[j], j) for j in 1:min(length(valtree), 3)])
+    y = pop!(valtree)
+    if !isempty(valtree) && i <= length(valtree)
+        @inbounds valtree[i] = y
+        _minmax_heap_trickle_down!(valtree, i)
+    end
+    return x    
+end
+
+function push!(h::BinaryMinMaxHeap, v)
+    valtree = h.valtree        
+    push!(valtree, v)
+    _minmax_heap_bubble_up!(valtree, length(valtree))
+end
+
+"""
+    top(h::BinaryMinMaxHeap)
+                        
+Get the top (minimum) of the heap.
+"""
+@inline top(h::BinaryMinMaxHeap) = min(h)
+
+@inline min(h::BinaryMinMaxHeap) = h.valtree[1]
+
+function max(h::BinaryMinMaxHeap) 
+    valtree = h.valtree
+    maxlen = min(length(valtree), 3)
+    @inbounds m = maximum([valtree[i] for i in 1:maxlen])
```

Comment:
aas of 1.0 you can just `return @inbounds...`

I think a nicer way to write this would be:

`return @inbounds maximum(valtree[1 : min(end, 3)])`